### ISSUE TITLE
Another CORS problem :-(

### ISSUE THREAD

**Issue by HJoost** [khaines/angular-applicationinsights#24]
Hi,

today I added angular-applicationinsights to our project. But it seems that I don't get data to the Insights Server because of a CORS problem:

XMLHttpRequest cannot load https://dc.services.visualstudio.com/v2/track. Request header field Access-Control-Allow-Headers is not allowed by Access-Control-Allow-Headers.

Fiddler Request Log:

OPTIONS https://dc.services.visualstudio.com/v2/track HTTP/1.1
Host: dc.services.visualstudio.com
Connection: keep-alive
Access-Control-Request-Method: POST
Origin: https://localhost:44302
User-Agent: Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.155 Safari/537.36
Access-Control-Request-Headers: accept, access-control-allow-headers, authorization, content-type
Accept: */*
Referer: https://localhost:44302/produktion/produktionsauftrag/48348
Accept-Encoding: gzip, deflate, sdch
Accept-Language: de-DE,de;q=0.8,en-US;q=0.6,en;q=0.4

Response: 
HTTP/1.1 200 OK
Cache-Control: no-cache
Pragma: no-cache
Expires: -1
Server: Microsoft-IIS/8.5
Access-Control-Allow-Methods: POST
Accept: application/json
Accept: application/x-json-stream
Set-Cookie: ai_session=4292f156...+00:00; expires=Fri, 21-Aug-2015 11:08:04 GMT; path=/
Set-Cookie: ai_user=f4b07e28f...+00:00; expires=Fri, 31-Dec-9999 23:59:59 GMT; path=/
X-Served-By: 9d1c395964..._Microsoft.Developer.Analytics.DataCollector_IN_4
X-Powered-By: ASP.NET
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept
Access-Control-Max-Age: 3600
Date: Fri, 21 Aug 2015 10:38:03 GMT
Content-Length: 0

Can anyone help me please?

**Comment by khaines** [comment]
The problem is that the request is trying to use headers "access-control-allow-headers" and "authorization". These are not allowed by the rules returned by the dc.services.visualstudio.com service. I'll have a look why these extra headers are in the options request (I'm assuming because they're in the original request), because they are not needed for AppInsight calls.

Are you using v0.2.5?

**Comment by HJoost** [comment]
I use the Files from this git repo (master branch). This angular-applicationinsights.js file is v0.2.4?

**Comment by HJoost** [comment]
Now it is partially working. There were two issues. In the app.config I had an entry, which I had to remove:
$httpProvider.defaults.headers.common['Access-Control-Allow-Headers'] = '*';

The other problem was that I set an default Authorization Header after login:
this.$http.defaults.headers.common['Authorization'] = 'Bearer ' + this.token;

But now I have to set this authorization header on every request I send to my WebApi. This is annoying. Does anyone know a workaround?

**Comment by khaines** [comment]
Ah, you were setting default headers. I will add code in the library to turn those off specifically when it's sending telemetry data. You shouldn't have to set those every time you want to use them :)

**Comment by HJoost** [comment]
Ok thank you!

I found a little help in setting a wrapper method for $http in my data services:

private callServer<T>(method: string, url: string, data: any): ng.IPromise<T>
        {
            var deferred = this.$q.defer();
            var req;
            if (data === undefined)
            {
                req = {
                    method: method,
                    url: this.webApiServer + url,
                    headers:
                    {
                        'Authorization': 'Bearer ' + this.token
                    }
                }
            } else
            {
                req = {
                    method: method,
                    url: this.webApiServer + url,
                    headers:
                    {
                        'Authorization': 'Bearer ' + this.token
                    },
                    data: data
                }
            }
            var result;
            this.$http(req)
                .then(response=>
                {
                    var result = <T>response.data;
                    deferred.resolve(result);
                })
                .catch(error=>
            {
                this.$log.error("Server Communication Error: ", error.status);
                if (error.status === 401)
                {
                    alert("Sie sind nicht eingeloggt oder haben keine Berechtigung diesen Inhalt zu laden");
                }
                throw error;
                deferred.reject();
                });
            return deferred.promise;
        }

**Comment by fushnisoft** [comment]
I had a similar issue recently, previously we had set 
`$httpProvider.defaults.headers.common['If-Modified-Since'] = '0'` to deal with IE caching issues but obviously this was a bit of a heavy solution. Knowing the specific calls that needed this allowed a more elegant solution of using an interceptor to only apply the header to specific urls. I am not sure you can hope to cleanup headers in the angular-applicationinsights module or if you even should?

_(typescript but hopefully you get the idea)_

```Typescript
module MyApp {
    'use strict';

    export interface IApiInterceptor {
        request: (config: angular.IRequestConfig) => angular.IRequestConfig;
    }

    export class ApiInterceptor implements IApiInterceptor {
        static serviceId: string = 'apiInterceptor';

        public request = (config: ng.IRequestConfig): angular.IRequestConfig => {
            config.headers = config.headers || {};
            // add headers to all api requests or relative URL requests (these are to our own server!)
            // We only want to add the headers to our own api calls because, for example, the app insights
            // server will reject requests with headers it doesn't like
            if (config.url.indexOf('api/') > 0 || config.url.indexOf('http') === -1) {
                // Disable IE request caching
                config.headers['If-Modified-Since'] = '0';
                // Glimpse support: http://anthonychu.ca/post/getting-angularjs-http-and-web-api-to-work-with-glimpse
                config.headers['X-Requested-With'] = 'XMLHttpRequest';
            }
            return config;
        }
    }

    angular
        .module('app')
        .factory(ApiInterceptor.serviceId, (): IApiInterceptor => { return new ApiInterceptor(); });

    angular
        .module('app')
        .config([
        '$httpProvider', ($httpProvider: angular.IHttpProvider): void => {
            $httpProvider.interceptors.push(ApiInterceptor.serviceId);
        }
    ]);
}
```

**Comment by HJoost** [comment]
Thank you for your reply! (And we also use Typescript, it's just perfect in conjunction with angular ;-))

**Comment by khaines** [comment]
" I am not sure you can hope to cleanup headers in the angular-applicationinsights module or if you even should?"

I've been thinking on this and it should be possible, and I think it is a good thing to do in the module (I'm open to opposing arguments). The http call that the app insights module is making is going to specifically the app insights data collection endpoint, so it is a "known" api from the module's perspective. Going on this basis I don't see a problem filtering the headers sent down to the ones that will be accepted via CORS, as long as it doesn't affect http calls outside of the module (ie: modifying the default config).

**Comment by sondreb** [comment]
I'm seeing the same issue, tried various workarounds with no luck. My app is secured with ADAL and JWT, not sure if that matters.

**Comment by khaines** [comment]
@sondreb I've uploaded a branch to this project (bugfix/khaines/cors-header-fix), where I've altered the tracking http call to remove all the headers that app insights won't accept. It should not interfere with your normal app interactions or security as the overrides only occur on the app insights call. So if you would be willing to do a verification of this fix, I would greatly appreciate a 2nd set of eyes. The /dist folder in that branch has the compiled/minified version of the corrected script.

**Comment by sondreb** [comment]
@khaines I got that branch, embedded the angular-applicationinsights.js into my concatenated lib.js (using Browserify) and it now throws this error in Chrome:

```
XMLHttpRequest cannot load https://dc.services.visualstudio.com/v2/track. Request header field Authorization is not allowed by Access-Control-Allow-Headers.
```

I have removed the  <meta http-equiv="Content-Security-Policy" temporarily, so it's not a CSP issue.

As mentioned, this project uses Azure Active Directory with the ADAL library to do authentication. It might inject stuff to the XHR requests, but I'm not sure.

When your library did not work, I changed to using another library (https://github.com/johnhidey/angular-appinsights) which is simply a small Angular provider wrapper around the official AppInsight library. This works fine in my solution, so you know that for reference. This loads the external library from Microsoft, which I would want to avoid as we're building an rich app that hopefully will have very few, if any, externally hosted dependencies.

Here are the HTTP headers that I get with the working code:

```
POST /v2/track HTTP/1.1
Host: dc.services.visualstudio.com
Connection: keep-alive
Content-Length: 1207
Pragma: no-cache
Cache-Control: no-cache
Origin: http://localhost:7601
User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/45.0.2454.101 Safari/537.36
Content-type: application/json
Accept: */*
DNT: 1
Referer: http://localhost:7601/
Accept-Encoding: gzip, deflate
Accept-Language: nb,en;q=0.8,en-US;q=0.6
```

I'm actually not seeing the request from angular-applicationinsights library, as I'm sure the Chrome browser is blocking the request since it knows it's about to submit auth header and the server doesn't allow it.

**Comment by khaines** [comment]
I'll make sure I uploaded the change to the branch properly... The code in that branch should be removing the auth header (and any other header) from the call to AI. The reason the base js sdk for app insights works is that it is using a raw XHR object. This is something I would like to avoid in the angular module since using the $http service allows for developers to mock in their own services during testing or at the very least prevent tests from sending telemetry. The $http object however has a default configuration setting which makes it easier for app developers to set common data (like auth headers) and this is what I'm trying to override for the AI call specifically. I'll double check the setup of that branch when I get home this evening and reply back to this thread.

**Comment by khaines** [comment]
In the non-minified code you are using, you should find a block of code like this:

```javascript
var headers = {};
for(var header in $http.defaults.headers.common){
	headers[header] = undefined;
}

for(var postHeader in $http.defaults.headers.post){
	headers[postHeader] = undefined;
}

headers.Accept = _contentType;
headers['Content-Type'] = _contentType;

var request = {
	method: 'POST',
	url:_analyticsServiceUrl,
	headers: headers,
	data:data,
	// bugfix for issue# 18: disable credentials on CORS requests.
	withCredentials: false 
};
```

This should be stripping out the default headers and making a custom headers object to use in the request. I did try it out previously on a test app that had a few headers like Authorization set, and they did get stripped out on my test app. But I was just running the default $http provider.

**Comment by khaines** [comment]
I double checked my fix vs what I think the cause of the issue is and the call is overriding the headers in common and post objects (& therefore sending the tracking call). The tracking call is a post so anything in the get/put/delete/etc default header objects should not affect it. However if you are still having the issue, I'm missing an obvious scenario and I would like to understand more about how your application is configuring the $http provider. If you would rather discuss this privately, you can find my contact information on my profile.

**Comment by Rogn** [comment]
@khaines We had the following code: `$http.defaults.headers.common = { ApiKey: window.getApiKey() }` and also saw the error talked about here. But using your fix above resolves the issue.

**Comment by khaines** [comment]
@Rogn : Thanks for replying and I'm glad that the code in the side branch fixed that for you. I will integrate this into the master branch for the next update.

**Comment by khaines** [comment]
fix added to v0.2.6 release
